### PR TITLE
Sync partner and projects with Mercury

### PR DESCRIPTION
**Pull request by MemunaHaruna**
- Adds partner and project models
- Listens to partner and project change events from Mercury and updates Scivias DB accordingly

#### How to test
- Ensure that Partner and Project pages can be viewed on the admin dashboard
- Verify that new partner and projects cannot be created via the admin dashboard UI

### CODE REVIEW

File: config/shoryuken.yml [hunk 1]
```diff
@@ -0,0 +1 @@
+  - ['<%= ENV['SQS_MERCURY_EVENT_QUEUE'] %>', 1]
```
Comment (dcassanego) [inline config/shoryuken.yml]: Doesn't this config file need more in it?  i see this in the equivalent Messaging API file:
```
aws:
  region: us-east-1
concurrency: 10
queues:
  - ['<%= ENV['SQS_MERCURY_EVENT_QUEUE'] %>', 1]
```

File: config/sqs_queues.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# Note that if you want Shoryuken to read from a queue, it needs to be set in shoryuken.yml
```
Comment (dcassanego) [inline config/sqs_queues.yml]: I don't see a place where this file is read or configured.

**Comment by MemunaHaruna** [comment]
Hi @dcassanego using the shoryuken config in Galapagos as a guide, I have removed the sqs_queues.yml and shoryuken.yml files.

Going through the shoryuken documentation [here](https://github.com/phstc/shoryuken/wiki/Getting-Started#create-a-queue), I was wondering if the queue will be created on Heroku via a rails console.

**Comment by MemunaHaruna** [comment]
Hi @dcassanego please take a look at the partner and project policy files I've added.

File: app/jobs/mercury_event_job.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  # }
+  def perform(_sqs_msg, command)
+    Rails.logger.info("MercuryEventJob :: Type `#{command['event_type']}` :: EventTime #{command['event_time']}")
+
+    case command['event_type']
+    when 'partner_update' then Partner.sync_from_mercury_event!(command['payload'])
+    else
+      raise ArgumentError, "MercuryEventJob :: Invalid event_type: #{command['event_type']}"
+    end
+  rescue Exception => e
```
Comment (dcassanego) [inline app/jobs/mercury_event_job.rb]: Don't rescue exception

Comment (dcassanego) [inline app/jobs/mercury_event_job.rb]: Avoid rescuing the `Exception` class. Perhaps you meant to rescue `StandardError`?

File: app/models/partner.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  #   data_lake_region: 'us-east',
+  #   created_at: '2013-05-01',
+  #   updated_at: '2020-01-21'
+  #   projects: [
+  #     {project_code: 'MC_BANCOLOMBIA', timezone_name: 'Bogota', active: 'true', created_at: '2020-01-21', updated_at: '2020-01-21'},
+  #     ...
+  #   ]
+  # }
+
+  def self.sync_from_mercury_event!(params)
```
Comment (dcassanego) [inline app/models/partner.rb]: Remove unused methods (partner#.)

Comment (dcassanego) [inline app/models/partner.rb]: Assignment Branch Condition size for sync_from_mercury_event! is too high. [<5, 17, 0> 17.72/15]

File: app/admin/projects.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
+ActiveAdmin.register Project do
+  config.clear_action_items!
+  actions :index, :show
+
+  menu parent: 'Partners', priority: 3
+  includes :partner
+  permit_params :partner_id, :project_code, :timezone_name,
+
```
Comment (dcassanego) [inline app/admin/projects.rb]: Empty line detected around arguments.

File: app/admin/projects.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  index do
```
Comment (dcassanego) [inline app/admin/projects.rb]: Align the arguments of a method call if they span more than one line.

File: app/models/partner.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
+class Partner < ApplicationRecord
+  has_many :projects
```
Comment (dcassanego) [inline app/models/partner.rb]: Specify a `:dependent` option.

File: app/models/partner.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  validates :name, :partner_code, presence: true
+  validates_uniqueness_of :partner_code
```
Comment (dcassanego) [inline app/models/partner.rb]: Prefer the new style validations `validates :column, uniqueness: value` over `validates_uniqueness_of`.

File: app/models/partner.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  #   projects: [
+  #     {project_code: 'MC_BANCOLOMBIA', timezone_name: 'Bogota', active: 'true', created_at: '2020-01-21', updated_at: '2020-01-21'},
+  #     ...
+  #   ]
+  # }
+
+  def self.sync_from_mercury_event!(params)
+    partner_params = params.with_indifferent_access.slice(:name, :partner_code, :updated_at)
+    partner = Partner
+              .where(partner_code: partner_params[:partner_code])
```
Comment (dcassanego) [inline app/models/partner.rb]: Place the . on the previous line, together with the method call receiver.

Comment (dcassanego) [inline app/models/partner.rb]: Use 2 (not 10) spaces for indenting an expression in an assignment spanning multiple lines.

File: app/models/partner.rb [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+              .first_or_create!(partner_params)
```
Comment (dcassanego) [inline app/models/partner.rb]: Place the . on the previous line, together with the method call receiver.

Comment (dcassanego) [inline app/models/partner.rb]: Use 2 (not 10) spaces for indenting an expression in an assignment spanning multiple lines.

File: app/models/partner.rb [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+    partner.update(partner_params)
+
+    params.with_indifferent_access['projects'].each do |project_params|
+      project_params = project_params.with_indifferent_access.slice(:project_code, :timezone_name, :updated_at)
+      project_params[:partner_id] = partner.id
+      project_params[:name] = project_params[:project_code] # To be manually updated later
+
+      Project
+      .where(project_code: project_params[:project_code])
```
Comment (dcassanego) [inline app/models/partner.rb]: Place the . on the previous line, together with the method call receiver.

Comment (dcassanego) [inline app/models/partner.rb]: Use 2 (not 0) spaces for indenting an expression spanning multiple lines.

File: app/models/partner.rb [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+      .first_or_create!(project_params)
```
Comment (dcassanego) [inline app/models/partner.rb]: Place the . on the previous line, together with the method call receiver.

Comment (dcassanego) [inline app/models/partner.rb]: Use 2 (not 0) spaces for indenting an expression spanning multiple lines.

File: app/models/partner.rb [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+      .update(project_params)
```
Comment (dcassanego) [inline app/models/partner.rb]: Place the . on the previous line, together with the method call receiver.

Comment (dcassanego) [inline app/models/partner.rb]: Use 2 (not 0) spaces for indenting an expression spanning multiple lines.

File: test/models/partner_test.rb [hunk 14]
```diff
@@ -0,0 +1,64 @@
+require 'test_helper'
+
+class PartnerTest < ActiveSupport::TestCase
+  it "creates a unique Partner" do
```
Comment (dcassanego) [inline test/models/partner_test.rb]: Prefer single-quoted strings when you don't need string interpolation or special symbols.

File: test/models/partner_test.rb [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    partner_1 = Partner.create!(
```
Comment (dcassanego) [inline test/models/partner_test.rb]: Use normalcase for variable numbers.

File: test/models/partner_test.rb [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+      name: 'Bancolombia',
+      partner_code: 'TEST'
```
Comment (dcassanego) [inline test/models/partner_test.rb]: Put a comma after the last parameter of a multiline method call.

File: test/models/partner_test.rb [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+    )
+    assert partner_1.valid?
+
+    partner_2 = Partner.new(
```
Comment (dcassanego) [inline test/models/partner_test.rb]: Use normalcase for variable numbers.

File: test/models/partner_test.rb [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+      name: 'Bancolombia',
+      partner_code: 'TEST'
```
Comment (dcassanego) [inline test/models/partner_test.rb]: Put a comma after the last parameter of a multiline method call.

File: test/models/partner_test.rb [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+    )
+    refute partner_2.valid?
```
Comment (dcassanego) [inline test/models/partner_test.rb]: Prefer `assert_not` over `refute`.

File: test/models/partner_test.rb [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+  end
+
+  test 'Sync from Mercury event' do
+    # First time, it should create new records
+    assert_difference 'Partner.count', 1 do
+      assert_difference 'Project.count', 1 do
+        Partner.sync_from_mercury_event!(partner_update_event('Bancolombia'))
+
+        partner = Partner.find_by_partner_code('BANCOLOMBIA')
```
Comment (dcassanego) [inline test/models/partner_test.rb]: Use `find_by` instead of dynamic `find_by_partner_code`.

File: test/models/partner_test.rb [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+      end
+    end
+
+    # Next time, it shouldn't create a new record, but it should update the name
+    assert_difference 'Partner.count', 0 do
+      assert_difference 'Project.count', 0 do
+        new_partner_name = 'New Brandname'
+        Partner.sync_from_mercury_event!(partner_update_event(new_partner_name))
+
+        assert_equal new_partner_name, Partner.find_by_partner_code('BANCOLOMBIA').name
```
Comment (dcassanego) [inline test/models/partner_test.rb]: Use `find_by` instead of dynamic `find_by_partner_code`.

File: test/models/partner_test.rb [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+    {
+      name: partner_name,
+      country: 'Colombia',
+      partner_code: 'BANCOLOMBIA',
+      pensieve_code: 'BANCOLOMBIA',
+      data_lake_region: 'us-east',
+      created_at: '2013-05-01',
+      updated_at: '2020-01-21',
+      projects: [
+        {project_code: 'MC_BANCOLOMBIA', timezone_name: 'Bogota', active: 'true', created_at: '2020-01-21', updated_at: '2020-01-21'},
```
Comment (dcassanego) [inline test/models/partner_test.rb]: Space inside { missing.

Comment (dcassanego) [inline test/models/partner_test.rb]: Space inside } missing.

File: test/models/partner_test.rb [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+      ]
```
Comment (dcassanego) [inline test/models/partner_test.rb]: Put a comma after the last item of a multiline hash.

File: test/models/project_test.rb [hunk 24]
```diff
@@ -0,0 +1,20 @@
+require "test_helper"
```
Comment (dcassanego) [inline test/models/project_test.rb]: Prefer single-quoted strings when you don't need string interpolation or special symbols.

File: test/models/project_test.rb [hunk 25]
```diff
@@ -0,0 +1,20 @@
...
+
+describe Project do
+  it "creates a unique Project" do
```
Comment (dcassanego) [inline test/models/project_test.rb]: Prefer single-quoted strings when you don't need string interpolation or special symbols.

File: test/models/project_test.rb [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
+    partner_id = partners(:one).id
+    project_1 = Project.create!(
```
Comment (dcassanego) [inline test/models/project_test.rb]: Use normalcase for variable numbers.

File: test/models/project_test.rb [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+      project_code: 'TEST_PROJECT',
+      timezone_name: 'Bogota',
+      partner_id: partner_id
```
Comment (dcassanego) [inline test/models/project_test.rb]: Put a comma after the last parameter of a multiline method call.

File: test/models/project_test.rb [hunk 28]
```diff
@@ -0,0 +1,20 @@
...
+    )
+    assert project_1.valid?
+
+    project_2 = Project.new(
```
Comment (dcassanego) [inline test/models/project_test.rb]: Use normalcase for variable numbers.

File: test/models/project_test.rb [hunk 29]
```diff
@@ -0,0 +1,20 @@
...
+      project_code: 'TEST_PROJECT',
+      timezone_name: 'Bogota',
+      partner_id: partner_id
```
Comment (dcassanego) [inline test/models/project_test.rb]: Put a comma after the last parameter of a multiline method call.

File: test/models/project_test.rb [hunk 30]
```diff
@@ -0,0 +1,20 @@
...
+    )
+    refute project_2.valid?
```
Comment (dcassanego) [inline test/models/project_test.rb]: Prefer `assert_not` over `refute`.

File: app/admin/projects.rb [hunk 31]
```diff
@@ -0,0 +1,34 @@
+ActiveAdmin.register Project do
```
Comment (dcassanego) [inline app/admin/projects.rb]: Block has too many lines. [28/25]

File: app/admin/projects.rb [hunk 32]
```diff
@@ -0,0 +1,34 @@
...
+  show do
+    attributes_table do
+      row :partner
+      row :name
+      row :project_code
+      row :timezone_name
+    end
+  end
+
+  form(decorate: true) do |f|
```
Comment (dcassanego) [inline app/admin/projects.rb]: Unused block argument - `f`. You can omit the argument if you don't care about it.

**Comment by MemunaHaruna** [comment]
Hi @cristina-gorrino I've added a name field to the Project table and also made that field editable in the admin UI since the name does not come from Mercury. When the project record is created from the Mercury event, it temporarily uses the Project code as its name. We can then update the names manually from the Scivias UI.

**Pull request merged**